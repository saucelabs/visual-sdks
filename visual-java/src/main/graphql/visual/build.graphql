query build($input: UUID!) {
  result: build(id: $input) {
    url
  }
}
