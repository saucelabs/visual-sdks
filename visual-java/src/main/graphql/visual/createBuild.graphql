mutation createBuild($input: BuildIn!) {
  result: createBuild(input: $input) {
    id
    name
    project
    branch
    status
    url
  }
}
