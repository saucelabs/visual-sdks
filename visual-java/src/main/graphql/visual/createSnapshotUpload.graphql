mutation createSnapshotUpload($input: SnapshotUploadIn!) {
  result: createSnapshotUpload(input: $input) {
    id
    buildId
    imageUploadUrl
  }
}
