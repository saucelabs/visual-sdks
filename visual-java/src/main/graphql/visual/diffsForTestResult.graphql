query diffsForTestResult($input: UUID!) {
  result: diffs(condition: { buildId: $input }) {
    nodes {
      id
      status
    }
  }
}
