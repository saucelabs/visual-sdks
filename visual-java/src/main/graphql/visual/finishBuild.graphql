mutation FinishBuild($input: FinishBuildIn!) {
  result: finishBuild(input: $input) {
    id
    name
    status
    url
  }
}
