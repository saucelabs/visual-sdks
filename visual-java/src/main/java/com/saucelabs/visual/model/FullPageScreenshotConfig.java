package com.saucelabs.visual.model;

import java.util.Arrays;
import java.util.List;
import org.openqa.selenium.WebElement;

public class FullPageScreenshotConfig {

  private int delayAfterScrollMs;
  private Boolean disableCSSAnimation;
  private List<String> hideAfterFirstScroll;
  private Boolean hideScrollBars;
  private int scrollLimit;
  private WebElement scrollElement;

  public FullPageScreenshotConfig(
      int delayAfterScrollMs,
      Boolean disableCSSAnimation,
      List<String> hideAfterFirstScroll,
      Boolean hideScrollBars,
      int scrollLimit,
      WebElement scrollElement) {
    this.delayAfterScrollMs = delayAfterScrollMs;
    this.disableCSSAnimation = disableCSSAnimation;
    this.hideAfterFirstScroll = hideAfterFirstScroll;
    this.hideScrollBars = hideScrollBars;
    this.scrollLimit = scrollLimit;
    this.scrollElement = scrollElement;
  }

  public static class Builder {
    private int delayAfterScrollMs;
    private Boolean disableCSSAnimation;
    private List<String> hideAfterFirstScroll;
    private Boolean hideScrollBars;
    private int scrollLimit;
    private WebElement scrollElement;

    public Builder withDelayAfterScrollMs(int delayAfterScrollMs) {
      this.delayAfterScrollMs = delayAfterScrollMs;
      return this;
    }

    public Builder withDisableCSSAnimation(Boolean disableCSSAnimation) {
      this.disableCSSAnimation = disableCSSAnimation;
      return this;
    }

    public Builder withHideAfterFirstScroll(String... hideAfterFirstScroll) {
      this.hideAfterFirstScroll = Arrays.asList(hideAfterFirstScroll);
      return this;
    }

    public Builder withHideScrollBars(Boolean hideScrollBars) {
      this.hideScrollBars = hideScrollBars;
      return this;
    }

    public Builder withScrollLimit(int scrollLimit) {
      this.scrollLimit = scrollLimit;
      return this;
    }

    public Builder withScrollElement(WebElement scrollElement) {
      this.scrollElement = scrollElement;
      return this;
    }

    public FullPageScreenshotConfig build() {
      return new FullPageScreenshotConfig(
          delayAfterScrollMs,
          disableCSSAnimation,
          hideAfterFirstScroll,
          hideScrollBars,
          scrollLimit,
          scrollElement);
    }
  }

  public int getDelayAfterScrollMs() {
    return delayAfterScrollMs;
  }

  public void setDelayAfterScrollMs(int delayAfterScrollMs) {
    this.delayAfterScrollMs = delayAfterScrollMs;
  }

  public Boolean getDisableCSSAnimation() {
    return disableCSSAnimation;
  }

  public void setDisableCSSAnimation(Boolean disableCSSAnimation) {
    this.disableCSSAnimation = disableCSSAnimation;
  }

  public List<String> getHideAfterFirstScroll() {
    return hideAfterFirstScroll;
  }

  public void setHideAfterFirstScroll(String... hideAfterFirstScroll) {
    this.hideAfterFirstScroll = Arrays.asList(hideAfterFirstScroll);
  }

  public Boolean getHideScrollBars() {
    return hideScrollBars;
  }

  public void setHideScrollBars(Boolean hideScrollBars) {
    this.hideScrollBars = hideScrollBars;
  }

  public int getScrollLimit() {
    return scrollLimit;
  }

  public void setScrollLimit(int scrollLimit) {
    this.scrollLimit = scrollLimit;
  }

  public WebElement getScrollElement() {
    return scrollElement;
  }

  public void setScrollElement(WebElement scrollElement) {
    this.scrollElement = scrollElement;
  }
}
