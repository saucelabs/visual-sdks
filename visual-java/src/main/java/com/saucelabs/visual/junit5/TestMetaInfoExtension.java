package com.saucelabs.visual.junit5;

import com.saucelabs.visual.TestMetaInfo;
import java.util.Optional;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
import org.junit.jupiter.api.extension.ExtensionContext;

public class TestMetaInfoExtension
    implements BeforeTestExecutionCallback, AfterTestExecutionCallback {
  @Override
  public void beforeTestExecution(ExtensionContext context) throws Exception {
    String className = context.getRequiredTestClass().getName();
    String testName = context.getRequiredTestMethod().getName();

    TestMetaInfo.THREAD_LOCAL.set(Optional.of(new TestMetaInfo(className, testName)));
  }

  @Override
  public void afterTestExecution(ExtensionContext context) throws Exception {
    TestMetaInfo.THREAD_LOCAL.set(Optional.empty());
  }
}
