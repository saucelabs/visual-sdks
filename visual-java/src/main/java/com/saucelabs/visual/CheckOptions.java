package com.saucelabs.visual;

import com.saucelabs.visual.graphql.type.DiffingOptionsIn;
import com.saucelabs.visual.model.DiffingFlag;
import com.saucelabs.visual.model.FullPageScreenshotConfig;
import com.saucelabs.visual.model.IgnoreRegion;
import com.saucelabs.visual.model.VisualRegion;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import org.openqa.selenium.WebElement;

public class CheckOptions {

  public enum DiffingMethod {
    SIMPLE,
    EXPERIMENTAL,
    BALANCED,
  }

  public CheckOptions() {}

  public CheckOptions(
      List<WebElement> ignoreElements,
      List<IgnoreRegion> ignoreRegions,
      List<VisualRegion> regions,
      String testName,
      String suiteName,
      DiffingMethod diffingMethod,
      DiffingOptionsIn diffingOptions,
      Boolean captureDom,
      String clipSelector,
      FullPageScreenshotConfig fullPageScreenshotConfig) {
    this.ignoreElements = ignoreElements;
    this.ignoreRegions = ignoreRegions;
    this.regions = regions;
    this.testName = testName;
    this.suiteName = suiteName;
    this.diffingMethod = diffingMethod;
    this.captureDom = captureDom;
    this.clipSelector = clipSelector;
    this.fullPageScreenshotConfig = fullPageScreenshotConfig;
    this.diffingOptions = diffingOptions;
  }

  private List<WebElement> ignoreElements = new ArrayList<>();
  private List<IgnoreRegion> ignoreRegions = new ArrayList<>();
  private List<VisualRegion> regions = new ArrayList<>();

  private String testName;
  private String suiteName;
  private DiffingMethod diffingMethod;
  private DiffingOptionsIn diffingOptions;
  private Boolean captureDom;
  private String clipSelector;

  private FullPageScreenshotConfig fullPageScreenshotConfig;

  public static class Builder {
    private List<WebElement> ignoreElements = new ArrayList<>();
    private List<IgnoreRegion> ignoreRegions = new ArrayList<>();
    private List<VisualRegion> regions = new ArrayList<>();
    private String testName;
    private String suiteName;
    private DiffingMethod diffingMethod;
    private DiffingOptionsIn diffingOptions;
    private Boolean captureDom;
    private String clipSelector;
    private FullPageScreenshotConfig fullPageScreenshotConfig;

    public Builder withIgnoreElements(List<WebElement> ignoreElements) {
      this.ignoreElements = ignoreElements;
      return this;
    }

    public Builder withIgnoreRegions(List<IgnoreRegion> ignoreRegions) {
      this.ignoreRegions = ignoreRegions;
      return this;
    }

    public Builder withTestName(String testName) {
      this.testName = testName;
      return this;
    }

    public Builder withSuiteName(String suiteName) {
      this.suiteName = suiteName;
      return this;
    }

    public Builder withDiffingMethod(DiffingMethod diffingMethod) {
      this.diffingMethod = diffingMethod;
      return this;
    }

    public Builder withCaptureDom(Boolean captureDom) {
      this.captureDom = captureDom;
      return this;
    }

    public Builder withClipSelector(String clipSelector) {
      this.clipSelector = clipSelector;
      return this;
    }

    public Builder withFullPageConfig(FullPageScreenshotConfig fullPageScreenshotConfig) {
      this.fullPageScreenshotConfig = fullPageScreenshotConfig;
      return this;
    }

    public Builder disableOnly(EnumSet<DiffingFlag> flags) {
      this.diffingOptions = new DiffingOptionsIn();
      DiffingFlag.setAll(this.diffingOptions, true);
      for (DiffingFlag f : flags) f.apply(this.diffingOptions, false);
      return this;
    }

    public Builder enableOnly(EnumSet<DiffingFlag> flags) {
      this.diffingOptions = new DiffingOptionsIn();
      DiffingFlag.setAll(this.diffingOptions, false);
      for (DiffingFlag f : flags) f.apply(this.diffingOptions, true);
      return this;
    }

    public Builder enableOnly(EnumSet<DiffingFlag> flags, WebElement element) {

      this.regions.add(VisualRegion.ignoreChangesFor(element).except(flags));
      return this;
    }

    public Builder disableOnly(EnumSet<DiffingFlag> flags, WebElement element) {

      this.regions.add(VisualRegion.detectChangesFor(element).except(flags));
      return this;
    }

    public CheckOptions build() {
      return new CheckOptions(
          ignoreElements,
          ignoreRegions,
          regions,
          testName,
          suiteName,
          diffingMethod,
          diffingOptions,
          captureDom,
          clipSelector,
          fullPageScreenshotConfig);
    }
  }

  public List<WebElement> getIgnoreElements() {
    return ignoreElements;
  }

  public void setIgnoreElements(List<WebElement> ignoreElements) {
    this.ignoreElements = ignoreElements;
  }

  public List<IgnoreRegion> getIgnoreRegions() {
    return ignoreRegions;
  }

  public void setIgnoreRegions(List<IgnoreRegion> ignoreRegions) {
    this.ignoreRegions = ignoreRegions;
  }

  public String getTestName() {
    return testName;
  }

  public void setTestName(String testName) {
    this.testName = testName;
  }

  public List<VisualRegion> getRegions() {
    return regions;
  }

  public void setRegions(List<VisualRegion> regions) {
    this.regions = regions;
  }

  public String getSuiteName() {
    return suiteName;
  }

  public void setSuiteName(String suiteName) {
    this.suiteName = suiteName;
  }

  public void setDiffingMethod(DiffingMethod diffingMethod) {
    this.diffingMethod = diffingMethod;
  }

  public DiffingMethod getDiffingMethod() {
    return diffingMethod;
  }

  public DiffingOptionsIn getDiffingOptions() {
    return diffingOptions;
  }

  public void setCaptureDom(Boolean captureDom) {
    this.captureDom = captureDom;
  }

  public Boolean getCaptureDom() {
    return captureDom;
  }

  public String getClipSelector() {
    return clipSelector;
  }

  public void setClipSelector(String clipSelector) {
    this.clipSelector = clipSelector;
  }

  public FullPageScreenshotConfig getFullPageScreenshotConfig() {
    return fullPageScreenshotConfig;
  }

  public void enableFullPageScreenshots(FullPageScreenshotConfig fullPageScreenshotConfig) {
    this.fullPageScreenshotConfig = fullPageScreenshotConfig;
  }

  public void enableFullPageScreenshots() {
    this.fullPageScreenshotConfig = new FullPageScreenshotConfig.Builder().build();
  }
}
