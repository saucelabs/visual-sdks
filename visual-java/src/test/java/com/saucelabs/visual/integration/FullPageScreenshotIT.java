package com.saucelabs.visual.integration;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.saucelabs.visual.CheckOptions;
import com.saucelabs.visual.junit5.TestMetaInfoExtension;
import com.saucelabs.visual.model.FullPageScreenshotConfig;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith({TestMetaInfoExtension.class, SnapshotExtension.class})
public class FullPageScreenshotIT extends IntegrationBase {
  Expect expect;

  @BeforeAll
  public static void login() {
    LoginPage loginPage = new LoginPage();
    loginPage.open();
    loginPage.login();
    InventoryLongPage inventoryPage = new InventoryLongPage();
    inventoryPage.open();
  }

  @Test
  public void takeAFPSWithDefaultSettings() {
    String id =
        sauceVisualCheck(
            "FPS",
            new CheckOptions.Builder().withFullPageConfig(new FullPageScreenshotConfig()).build());
    String result = getSnapshotResult(id);
    expect.toMatchSnapshot(result);
  }

  @Test
  public void takeAFPSAndDisplayScrollBars() {
    String id =
        sauceVisualCheck(
            "FPS hideScrollBars=false",
            new CheckOptions.Builder()
                .withHideScrollBars(false)
                .withFullPageConfig(new FullPageScreenshotConfig())
                .build());
    String result = getSnapshotResult(id);
    expect.toMatchSnapshot(result);
  }

  @Test
  public void takeAFPSWithScrollLimit() {
    String id =
        sauceVisualCheck(
            "FPS scrollLimit=1",
            new CheckOptions.Builder()
                .withHideScrollBars(true)
                .withFullPageConfig(new FullPageScreenshotConfig())
                .build());
    String result = getSnapshotResult(id);
    expect.toMatchSnapshot(result);
  }
}
