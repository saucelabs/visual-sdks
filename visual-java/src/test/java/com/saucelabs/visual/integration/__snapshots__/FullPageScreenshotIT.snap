com.saucelabs.visual.integration.FullPageScreenshotIT.takeAFPSAndDisplayScrollBars=[
{
  "snapshot" : {
    "name" : "FPS hideScrollBars=false (Sauce)",
    "suiteName" : "com.saucelabs.visual.integration.FullPageScreenshotIT",
    "testName" : "takeAFPSAndDisplayScrollBars",
    "operatingSystem" : "WINDOWS",
    "operatingSystemVersion" : "10",
    "device" : "Desktop (1384x1007)",
    "devicePixelRatio" : 1,
    "hasDom" : true,
    "diffs" : {
      "nodes" : [ {
        "status" : "EQUAL",
        "diffClusters" : [ ]
      } ]
    },
    "ignoreRegions" : [ ]
  }
}
]


com.saucelabs.visual.integration.FullPageScreenshotIT.takeAFPSWithDefaultSettings=[
{
  "snapshot" : {
    "name" : "FPS (Sauce)",
    "suiteName" : "com.saucelabs.visual.integration.FullPageScreenshotIT",
    "testName" : "takeAFPSWithDefaultSettings",
    "operatingSystem" : "WINDOWS",
    "operatingSystemVersion" : "10",
    "device" : "Desktop (1384x1007)",
    "devicePixelRatio" : 1,
    "hasDom" : true,
    "diffs" : {
      "nodes" : [ {
        "status" : "EQUAL",
        "diffClusters" : [ ]
      } ]
    },
    "ignoreRegions" : [ ]
  }
}
]


com.saucelabs.visual.integration.FullPageScreenshotIT.takeAFPSWithScrollLimit=[
{
  "snapshot" : {
    "name" : "FPS scrollLimit=1 (Sauce)",
    "suiteName" : "com.saucelabs.visual.integration.FullPageScreenshotIT",
    "testName" : "takeAFPSWithScrollLimit",
    "operatingSystem" : "WINDOWS",
    "operatingSystemVersion" : "10",
    "device" : "Desktop (1384x1007)",
    "devicePixelRatio" : 1,
    "hasDom" : true,
    "diffs" : {
      "nodes" : [ {
        "status" : "EQUAL",
        "diffClusters" : [ ]
      } ]
    },
    "ignoreRegions" : [ ]
  }
}
]